File: benchmarks/complex/btree/btree.rb
ORDER = 64
MAX_KEYS = ORDER - 1

Node = Struct.new(:keys, :children, :n, :is_leaf)

$nodes = []

def new_node(leaf)
  nd = Node.new(Array.new(MAX_KEYS, 0), Array.new(ORDER, -1), 0, leaf)
  $nodes << nd
  $nodes.length - 1
end

def search(nid, key)
  return false if nid < 0
  nd = $nodes[nid]
  i = 0
  while i < nd.n && key > nd.keys[i]; i += 1; end
  return true if i < nd.n && key == nd.keys[i]
  return false if nd.is_leaf
  search(nd.children[i], key)
end

def split_child(parent, idx)
  full = $nodes[parent].children[idx]
  mid = MAX_KEYS / 2
  right = new_node($nodes[full].is_leaf)
  rn = MAX_KEYS - mid - 1
  rn.times { |j| $nodes[right].keys[j] = $nodes[full].keys[mid + 1 + j] }
  unless $nodes[full].is_leaf
    (rn + 1).times { |j| $nodes[right].children[j] = $nodes[full].children[mid + 1 + j] }
  end
  promote_key = $nodes[full].keys[mid]
  $nodes[full].n = mid
  $nodes[right].n = rn
  pn = $nodes[parent].n
  j = pn
  while j > idx
    $nodes[parent].keys[j] = $nodes[parent].keys[j-1]
    $nodes[parent].children[j+1] = $nodes[parent].children[j]
    j -= 1
  end
  $nodes[parent].keys[idx] = promote_key
  $nodes[parent].children[idx + 1] = right
  $nodes[parent].n += 1
end

def insert_nonfull(nid, key)
  nd = $nodes[nid]
  if nd.is_leaf
    i = nd.n - 1
    while i >= 0 && key < nd.keys[i]
      nd.keys[i+1] = nd.keys[i]; i -= 1
    end
    return if i >= 0 && nd.keys[i] == key
    nd.keys[i+1] = key; nd.n += 1
  else
    i = nd.n - 1
    while i >= 0 && key < nd.keys[i]; i -= 1; end
    return if i >= 0 && nd.keys[i] == key
    i += 1
    ci = nd.children[i]
    if $nodes[ci].n == MAX_KEYS
      split_child(nid, i)
      i += 1 if key > $nodes[nid].keys[i]
      return if i < $nodes[nid].n && key == $nodes[nid].keys[i]
    end
    insert_nonfull($nodes[nid].children[i], key)
  end
end

$root = new_node(true)

def insert(key)
  if $nodes[$root].n == MAX_KEYS
    s = new_node(false)
    $nodes[s].children[0] = $root
    $root = s
    split_child(s, 0)
    insert_nonfull(s, key)
  else
    insert_nonfull($root, key)
  end
end

$rng = 12345
def next_rand
  $rng = ($rng * 1103515245 + 12345) & 0x7FFFFFFF
  $rng
end

N = 1000000
t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)

$rng = 12345
N.times { insert(next_rand % (N * 10)) }

$rng = 12345
found = 0
N.times { found += 1 if search($root, next_rand % (N * 10)) }

$rng = 99999
found2 = 0
N.times { found2 += 1 if search($root, next_rand % (N * 10)) }

ms = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000).to_i
puts "inserted: #{N}"
puts "found (known): #{found}"
puts "found (random): #{found2}"
puts "nodes: #{$nodes.length}"
puts "time: #{ms} ms"

File: benchmarks/complex/btree/btree.c
/* B-tree benchmark — order-64 B-tree, insert 1M keys, query 1M, delete 500K
   Reports operation counts and timing */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define ORDER 64
#define MAX_KEYS (ORDER - 1)
#define MIN_KEYS (ORDER / 2 - 1)
#define MAX_NODES 100000

typedef struct {
    int keys[MAX_KEYS];
    int children[ORDER];
    int n;       /* number of keys */
    int is_leaf;
} Node;

static Node nodes[MAX_NODES];
static int node_cnt;
static int root;

static int new_node(int leaf) {
    int id = node_cnt++;
    nodes[id].n = 0;
    nodes[id].is_leaf = leaf;
    memset(nodes[id].children, -1, sizeof(nodes[id].children));
    return id;
}

static int search(int id, int key) {
    if (id < 0) return 0;
    Node *nd = &nodes[id];
    int i = 0;
    while (i < nd->n && key > nd->keys[i]) i++;
    if (i < nd->n && key == nd->keys[i]) return 1;
    if (nd->is_leaf) return 0;
    return search(nd->children[i], key);
}

static void split_child(int parent, int idx) {
    int full = nodes[parent].children[idx];
    int mid = MAX_KEYS / 2;
    int right = new_node(nodes[full].is_leaf);

    nodes[right].n = MAX_KEYS - mid - 1;
    for (int j = 0; j < nodes[right].n; j++)
        nodes[right].keys[j] = nodes[full].keys[mid + 1 + j];
    if (!nodes[full].is_leaf)
        for (int j = 0; j <= nodes[right].n; j++)
            nodes[right].children[j] = nodes[full].children[mid + 1 + j];

    int promote_key = nodes[full].keys[mid];
    nodes[full].n = mid;

    /* Shift parent keys/children right */
    for (int j = nodes[parent].n; j > idx; j--) {
        nodes[parent].keys[j] = nodes[parent].keys[j-1];
        nodes[parent].children[j+1] = nodes[parent].children[j];
    }
    nodes[parent].keys[idx] = promote_key;
    nodes[parent].children[idx + 1] = right;
    nodes[parent].n++;
}

static void insert_nonfull(int id, int key) {
    Node *nd = &nodes[id];
    int i = nd->n - 1;
    if (nd->is_leaf) {
        while (i >= 0 && key < nd->keys[i]) {
            nd->keys[i+1] = nd->keys[i];
            i--;
        }
        /* Skip duplicates */
        if (i >= 0 && nd->keys[i] == key) return;
        nd->keys[i+1] = key;
        nd->n++;
    } else {
        while (i >= 0 && key < nd->keys[i]) i--;
        if (i >= 0 && nd->keys[i] == key) return;
        i++;
        if (nodes[nd->children[i]].n == MAX_KEYS) {
            split_child(id, i);
            if (key > nd->keys[i]) i++;
            if (key == nd->keys[i]) return;
        }
        insert_nonfull(nd->children[i], key);
    }
}

static void insert(int key) {
    if (nodes[root].n == MAX_KEYS) {
        int s = new_node(0);
        nodes[s].children[0] = root;
        split_child(s, 0);
        root = s;
        insert_nonfull(s, key);
    } else {
        insert_nonfull(root, key);
    }
}

/* Simple LCG for deterministic pseudo-random numbers */
static long rng_state = 12345;
static int next_rand(void) {
    rng_state = (rng_state * 1103515245 + 12345) & 0x7FFFFFFF;
    return (int)rng_state;
}

int main(void) {
    int N = 1000000;
    node_cnt = 0;
    root = new_node(1);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    /* Insert */
    rng_state = 12345;
    for (int i = 0; i < N; i++) {
        int key = next_rand() % (N * 10);
        insert(key);
    }

    /* Query */
    rng_state = 12345;
    int found = 0;
    for (int i = 0; i < N; i++) {
        int key = next_rand() % (N * 10);
        if (search(root, key)) found++;
    }

    /* Query random (mix of hits and misses) */
    rng_state = 99999;
    int found2 = 0;
    for (int i = 0; i < N; i++) {
        int key = next_rand() % (N * 10);
        if (search(root, key)) found2++;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    long ms = (t1.tv_sec - t0.tv_sec)*1000 + (t1.tv_nsec - t0.tv_nsec)/1000000;
    printf("inserted: %d\n", N);
    printf("found (known): %d\n", found);
    printf("found (random): %d\n", found2);
    printf("nodes: %d\n", node_cnt);
    printf("time: %ld ms\n", ms);
    return 0;
}
